src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,392 @@
+/-
+Copyright (c) 2021 Gabriel Moise. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Moise
+-/
+
+import algebra.big_operators.basic
+import combinatorics.simple_graph.basic
+import data.fintype.basic
+import data.sym2
+import linear_algebra.matrix
+
+/-!
+# Incidence matrices
+
+This module defines the incidence matrix `inc_matrix` of an undirected graph `simple_graph`,
+and provides theorems and lemmas connecting graph properties to computational properties of the
+matrix. It also defines the notion of `orientation` for a `simple_graph`, picking a direction for
+each undirected edge in the graph and then defining the oriented incidence matrix
+`oriented_inc_matrix` based on that.
+
+## Main definitions
+
+* `inc_matrix` : the incidence matrix `M` of a `simple_graph` with coefficients in a given ring R.
+* `orientation` : a structure that defines a choice of direction on the edges of a `simple_graph`.
+* `oriented_inc_matrix` : the oriented incidence matrix `N(o)` of a `simple_graph` with
+respect to a given `orientation`.
+
+## Main results
+
+1. `adj_sum_of_mul_inc_one` : `∑ e : E, M i e * M j e = 1`, for any two adjacent vertices
+`i` and `j`.
+2. `inc_matrix_mul_non_adj` : `M i e * M j e = 0`, for any two distinct non-adjacent vertices
+`i`, `j` and edge `e`.
+3. `inc_matrix_element_power_id` : every element from `M` is idempotent.
+4. `degree_equals_sum_of_incidence_row` : for any vertex `i`, the sum on the ith row of `M` is
+equal to the degree of `i`.
+5. `oriented_inc_matrix_elem_squared` : `(N(o) i e) ^ 2 = M i e`, for any orientation `o`,
+vertex `i` and edge `e`.
+6. `oriented_inc_matrix_mul_of_adj` : for any adjacent vertices `i`, `j` and edge `e`,
+`N(o) i e * N(o) j e = if e = ⟦(i, j)⟧ then -1 else 0`.
+7. `oriented_inc_matrix_mul_non_adj` : for any non-adjacent distinct vertices `i`, `j` and
+edge `e`, `N(o) i e * N(o) j e = 0`.
+8. `vec_mul_oriented_inc_matrix` : `(xᵀ ⬝ N(o)) e = x o.head(e) - x o.tail(e).`
+
+## References
+
+<https://en.wikipedia.org/wiki/Orientation_(graph_theory)>
+-/
+
+open_locale big_operators matrix
+open finset matrix simple_graph sym2
+
+universes u v
+variables {R : Type u} {V : Type v}
+
+section to_move
+
+@[simp]
+lemma ite_mul_ite_zero_right
+  {R : Type*} [mul_zero_class R] {P Q : Prop} [decidable P] [decidable Q] (a b : R):
+  (ite P a 0) * (ite Q b 0) = ite (P ∧ Q) (a * b) 0 :=
+by simp [←ite_and]
+
+end to_move
+
+namespace simple_graph
+
+variables (G : simple_graph V)
+
+-- ### Helping lemmas for edges
+
+lemma edge_set_ne {u v : V} {e: G.edge_set} (h : ↑e = ⟦(u, v)⟧) : u ≠ v :=
```

Comment:
If you could use this form of it instead, it might be worth moving the lemma near `simple_graph.ne_of_adj`.
```
lemma ne_of_mem_edge_set {u v : V} : ⟦(u, v)⟧ ∈ G.edge_set → u ≠ v := G.ne_of_adj
```